### ISSUE TITLE
Discrepancy in between the loss terms mentioned in the paper and the code. 

### ISSUE THREAD

**Issue by bhattg** [jsyoon0823/TimeGAN#2]
Dear author, 

I have the following doubts related to the code :
- In the following lines of the code TimeGAN.py line 202: 
`  # 3. Two Momments
  G_loss_V1 = tf.reduce_mean(tf.abs(tf.sqrt(tf.nn.moments(X_hat,[0])[1] + 1e-6) - tf.sqrt(tf.nn.moments(X,[0])[1] + 1e-6)))
  G_loss_V2 = tf.reduce_mean(tf.abs((tf.nn.moments(X_hat,[0])[0]) - (tf.nn.moments(X,[0])[0])))`, 
To which term in the paper do these loss terms contribute?

- Line 280 of the same file, 
` if (check_d_loss > 0.15):        
      _, step_d_loss = sess.run([D_solver, D_loss], feed_dict={X: X_mb, T: T_mb, Z: Z_mb})`
Why this value (0.15) has to be same for all the datasets? 

- In the paper, it was mentioned that z_t are being sampled from the Wiener process. However, in the code, they are being sampled from a uniform distribution, which is not very common in the realm of time series. Any specific reasons? 

- In the definition of L_S, the way it is written is giving the impression that g_X is just being used for one-step dynamics since h_t's are already there from the autoencoder's encoder, in contrast to what is given in the code. So how does that equation capture long-range dynamics? 

Thanks for your time, and let me know if I am misconstruing something.

**Comment by jsyoon0823** [comment]
Thank you for your interest in our paper.

1. This is more like moment matching tricks in generative models.
- This is useful for increasing the variance of the generated samples (helpful for reducing mode collapse problems).

2. This is for avoiding dominant discriminator
- In GAN training, the discriminator is usually trained faster than the generator. 
- However, as you know, the balance between generator and discriminator training is critical in GAN training. 
- Therefore, we adjust the speed of discriminator training.
- 0.15 is somewhat reasonable number across all the datasets to avoid dominant discriminator.

3. Wiener vs Uniform
- We tried both Wiener and Uniform distributions.
- However, in terms of the performance, there is not much difference. Therefore, to simplify the code, we just use uniform distribution.
- If you want to use Wiener process instead of Uniform distribution, I think you can easily modify the codes of z_t generation part.

4. 
- To predict the one-step dynamics, the RNN should capture the long-term dynamics.
- Therefore, h_t and e_t should capture them well.
- Through h_t and e_t, TimeGAN can capture those long-range dynamics.

Thanks
